### PR TITLE
CLOUDSTACK-8996: Reducing Virtual Machine Deployments

### PR DESCRIPTION
**Pull request by meenakshim1**
CLOUDSTACK-8996: Updates testscript 'testpath_reduce_vm_deployments.py'

Reference to #1010

### CODE REVIEW

File: test/integration/testpaths/testpath_reduce_vm_deployments.py [hunk 1]
```diff
@@ -0,0 +1,2076 @@
...
+        cls.services = testClient.getParsedTestDataConfig()
+        cls.hypervisor = testClient.getHypervisorInfo()
+
+        # Get Zone, Domain and templates
+        domain = get_domain(cls.apiclient)
+        cls.zone = get_zone(cls.apiclient, cls.testClient.getZoneForTests())
+        cls.services['mode'] = cls.zone.networktype
+        cls.domain = get_domain(cls.apiclient)
+
+        cls.mgtSvrDetails = cls.config.__dict__["mgtSvr"][0].__dict__
```
Comment (pavanb018) [inline test/integration/testpaths/testpath_reduce_vm_deployments.py]: You are accessing private variables, do you need management server details some where in the file , I don't see any use ? Can you please let me know if i missed it somewhere ?

File: test/integration/testpaths/testpath_reduce_vm_deployments.py [hunk 2]
```diff
@@ -0,0 +1,2076 @@
...
+    @classmethod
+    def setUpClass(cls):
+        testClient = super(TestDeployVMandVMLifeCycle, cls).getClsTestClient()
+        cls.apiclient = testClient.getApiClient()
+        cls.services = testClient.getParsedTestDataConfig()
+        cls.hypervisor = testClient.getHypervisorInfo()
+
+        # Get Zone, Domain and templates
+        domain = get_domain(cls.apiclient)
+        cls.zone = get_zone(cls.apiclient, cls.testClient.getZoneForTests())
```
Comment (pavanb018) [inline test/integration/testpaths/testpath_reduce_vm_deployments.py]: Your testClient is not a cls variable , however in this function you have called cls.testClinet.getZoneForTests(). Either make upper one a cls variable or correct in function call

**Comment by meenakshim1** [comment]
**Test Execution Results**

Test to create service offering ... === TestName: test_01_create_service_offering | Status : SUCCESS ===
ok
Test deploy virtual machine with root resize ... === TestName: test_00_deploy_vm_root_resize | Status : SUCCESS ===
ok
Test proper failure to deploy virtual machine with ... === TestName: test_01_deploy_vm_root_resize | Status : SUCCESS ===
ok
Test proper failure to deploy virtual machine with ... === TestName: test_02_deploy_vm_root_resize | Status : SUCCESS ===
ok
Test Stop Virtual Machine ... SKIP: At least two hosts should be present in the zone for migration
Test to update existing service offering ... === TestName: test_02_edit_service_offering | Status : SUCCESS ===
ok
Test destroy(expunge) Virtual Machine ... === TestName: test_02_expunge_vm | Status : SUCCESS ===
ok
Test to delete service offering ... === TestName: test_03_delete_service_offering | Status : SUCCESS ===
ok
Test to change service to a small capacity ... === TestName: test_04_change_offering_small | Status : SUCCESS ===
ok
Test Deploy Virtual Machine ... === TestName: test_04_deploy_vm | Status : SUCCESS ===
ok
Test advanced zone virtual router ... === TestName: test_advZoneVirtualRouter | Status : SUCCESS ===
ok
Tests for basic zone virtual router ... === TestName: test_basicZoneVirtualRouter | Status : SUCCESS ===
ok
Test userdata as POST, size > 2k ... === TestName: test_deployvm_userdata_post | Status : SUCCESS ===
ok
test DeployVM in anti-affinity groups ... SKIP: At least two hosts should be present in the zone to verify affinity group of type host-antiaffinity
Test userdata as POST, size > 2k ... === TestName: test_deployvm_userdata_post | Status : SUCCESS ===
ok
Test Deploy Virtual Machine ... SKIP: Skipping test because suitable hypervisor/host not                    present

----------------------------------------------------------------------
Ran 16 tests in 1121.082s

OK (SKIP=3)

File: test/integration/testpaths/testpath_reduce_vm_deployments.py (hunk 1 above)
Comment (meenakshim1) [inline test/integration/testpaths/testpath_reduce_vm_deployments.py]: Updated

File: test/integration/testpaths/testpath_reduce_vm_deployments.py (hunk 2 above)
Comment (meenakshim1) [inline test/integration/testpaths/testpath_reduce_vm_deployments.py]: Updated

File: test/integration/testpaths/testpath_reduce_vm_deployments.py [hunk 3]
```diff
@@ -0,0 +1,2073 @@
...
+        self.debug("Scaling VM-ID: %s to service offering: %s and state %s" % (
+            self.virtual_machine.id,
+            self.big_offering.id,
+            self.virtual_machine.state
+        ))
+
+        cmd = scaleVirtualMachine.scaleVirtualMachineCmd()
+        cmd.serviceofferingid = self.big_offering.id
+        cmd.id = self.virtual_machine.id
+        self.apiclient.scaleVirtualMachine(cmd)
```
Comment (shwetaag) [inline test/integration/testpaths/testpath_reduce_vm_deployments.py]: use scale function from vm class and if its not there better create it there

File: test/integration/testpaths/testpath_reduce_vm_deployments.py [hunk 4]
```diff
@@ -0,0 +1,2073 @@
...
+        """Test destroy(expunge) Virtual Machine
+        """
+        # Validate the following
+        # 1. listVM command should NOT  return this VM any more.
+
+        self.debug("Expunge VM-ID: %s" % self.small_virtual_machine.id)
+
+        cmd = destroyVirtualMachine.destroyVirtualMachineCmd()
+        cmd.id = self.small_virtual_machine.id
+        self.apiclient.destroyVirtualMachine(cmd)
```
Comment (shwetaag) [inline test/integration/testpaths/testpath_reduce_vm_deployments.py]: Use destroy lib function from virtual machine base class

File: test/integration/testpaths/testpath_reduce_vm_deployments.py [hunk 5]
```diff
@@ -0,0 +1,2073 @@
...
+            command = "umount %s" % mount_dir
+            ssh_client.execute(command)
+        except Exception as e:
+            self.fail("SSH failed for virtual machine: %s - %s" %
+                      (self.virtual_machine.ipaddress, e))
+
+        # Detach from VM
+        cmd = detachIso.detachIsoCmd()
+        cmd.virtualmachineid = self.virtual_machine.id
+        self.apiclient.detachIso(cmd)
```
Comment (shwetaag) [inline test/integration/testpaths/testpath_reduce_vm_deployments.py]: again use library functions for detach iso .. Its always a good idea to no call api commands from test cases

File: test/integration/testpaths/testpath_reduce_vm_deployments.py [hunk 6]
```diff
@@ -0,0 +1,2073 @@
...
+        #    a valid information for newly created offering
+
+        # Generate new name & displaytext from random data
+        random_displaytext = random_gen()
+        random_name = random_gen()
+
+        self.debug("Updating service offering with ID: %s" %
+                   self.service_offering_1.id)
+
+        cmd = updateServiceOffering.updateServiceOfferingCmd()
```
Comment (shwetaag) [inline test/integration/testpaths/testpath_reduce_vm_deployments.py]: move this to library function

File: test/integration/testpaths/testpath_reduce_vm_deployments.py [hunk 7]
```diff
@@ -0,0 +1,2073 @@
...
+            self.skipTest(
+                "Skipping this test for {} due to"
+                " bug CS-38153".format(self.hypervisor))
+        try:
+            self.medium_virtual_machine.stop(self.apiclient)
+        except Exception as e:
+            self.fail("Failed to stop VM: %s" % e)
+
+        cmd = \
+            changeServiceForVirtualMachine.changeServiceForVirtualMachineCmd()
```
Comment (shwetaag) [inline test/integration/testpaths/testpath_reduce_vm_deployments.py]: move this api to library

File: test/integration/testpaths/testpath_reduce_vm_deployments.py [hunk 8]
```diff
@@ -0,0 +1,2073 @@
...
+
+class TestData(object):
+
+    """Test data object that is required to create resources
+    """
+
+    def __init__(self):
+        self.testdata = {
+            # data to create an account
+            "account": {
```
Comment (shwetaag) [inline test/integration/testpaths/testpath_reduce_vm_deployments.py]: move entire test data to marvin test data file

File: test/integration/testpaths/testpath_reduce_vm_deployments.py [hunk 9]
```diff
@@ -0,0 +1,2073 @@
...
+        except Exception as e:
+            self.fail("Exception while downloading ISO %s: %s"
+                      % (iso.id, e))
+
+        self.debug("Attach ISO with ID: %s to VM ID: %s" % (
+            iso.id,
+            self.virtual_machine.id
+        ))
+        # Attach ISO to virtual machine
+        cmd = attachIso.attachIsoCmd()
```
Comment (shwetaag) [inline test/integration/testpaths/testpath_reduce_vm_deployments.py]: move this also to library

File: test/integration/testpaths/testpath_reduce_vm_deployments.py (hunk 3 above)
Comment (meenakshim1) [inline test/integration/testpaths/testpath_reduce_vm_deployments.py]: Updated

File: test/integration/testpaths/testpath_reduce_vm_deployments.py (hunk 4 above)
Comment (meenakshim1) [inline test/integration/testpaths/testpath_reduce_vm_deployments.py]: Updated.

File: test/integration/testpaths/testpath_reduce_vm_deployments.py (hunk 5 above)
Comment (meenakshim1) [inline test/integration/testpaths/testpath_reduce_vm_deployments.py]: Updated the script by using lib function from base.py

File: test/integration/testpaths/testpath_reduce_vm_deployments.py (hunk 6 above)
Comment (meenakshim1) [inline test/integration/testpaths/testpath_reduce_vm_deployments.py]: Added updatedServiceOffering method to base.py , and refactored the script accordingly.

File: test/integration/testpaths/testpath_reduce_vm_deployments.py (hunk 7 above)
Comment (meenakshim1) [inline test/integration/testpaths/testpath_reduce_vm_deployments.py]: Done

File: test/integration/testpaths/testpath_reduce_vm_deployments.py (hunk 8 above)
Comment (meenakshim1) [inline test/integration/testpaths/testpath_reduce_vm_deployments.py]: Done

File: test/integration/testpaths/testpath_reduce_vm_deployments.py (hunk 9 above)
Comment (meenakshim1) [inline test/integration/testpaths/testpath_reduce_vm_deployments.py]: Done

**Comment by meenakshim1** [comment]
Redundant Test.

**Pull request closed without merging**